scripts/g.extension/g.extension.py

```diff
@@ -207,14 +208,210 @@
     MAKE = "make"
 
 
+class GitAdapter:
+    """
+    Basic class for listing and downloading GRASS GIS AddOns using git
+
+    """
+
+    def __init__(
+        self,
+        url="https://github.com/osgeo/grass-addons",
+        working_directory=None,
+        official_repository_structure=True,
+        major_grass_version=8,
+        branch=None,
+        verbose=False,
+        quiet=False,
+    ):
+        #: Attribute containg the URL to the online repository
+        self.url = url
+        self.major_grass_version = major_grass_version
+        #: Attribute flagging if the repository is structured like the official addons repository
+        self.official_repository_structure = official_repository_structure
+        #: Attribute containg the path to the working directory where the repo is cloned out to
+        if working_directory:
+            self.working_directory = Path(working_directory).absolute()
+        else:
+            self.working_directory = Path().absolute()
+
+        # Check if working directory is writable
+        self.__check_permissions()
+
+        #: Attribute containg available branches
+        self.branches = self._get_branch_list()
+        # Initialize the local copy
+        self._initialize_clone()
+        #: Attribute containg the default branch of the repository
+        self.default_branch = self._get_default_branch()
+        #: Attribute containg the branch used for checkout
+        self.branch = self._set_branch(branch)
+        #: Attribute containg list of addons in the repository with path to directories
+        self.addons = self._get_addons_list()
+
+    def _initialize_clone(self):
+        """Get a minimal working copy of a git repository without content"""
+        if not self.working_directory.exists():
+            self.working_directory.mkdir(exist_ok=True, parents=True)
+        gs.call(
+            ["git", "clone", "--no-checkout", "--depth=1", "--filter=tree:0", self.url],
+            cwd=self.working_directory,
+        )
+        repo_directory = (
+            self.url.split("/")[-1][0:-4]
+            if self.url.endswith(".git")
+            else self.url.split("/")[-1]
```

Comment:
I'm not sure about best practice for splitting URLs (split vs Path vs urlparse), but rsplit with max splits will be better than split.